### ISSUE TITLE
MaxAutoRenewDuration is not closing the lock threads

### ISSUE THREAD

**Issue by gereyhe** [Azure/azure-sdk-for-net#11213]
The 'MaxAutoRenewDuration = TimeSpan.FromMinutes(300)' is creating a thread to do a lock renewal. By default the library code does the renewal, when the messages are consumed at the very first second they arrived in the backend, the thread created during the default <5min is closed, however due to an unknown reason the code is not closing the MaxAutoRenewDuration thread for all the events that are incoming, as a result, the following error is displayed:

'The  lock supplied is invalid. Either the lock expired, or the message has already been removed from the queue, or was received by a different receiver instance.'

The sequence is the following:

1. Message arrive to the queue in 01-Apr-20 at 02:55:24
2. Message is consumed in the same second it arrived in 01-Apr-20 at 02:55:24
 
After this point the message has arrived into the queue, locked with the LockToken a51cba32-8fd7-43c7-b4be-177099e14884 to give it to 1 consumer and then processed and consumed in the same second 24, which is a complete call. After that this is what is happening:
 
3. After 5 minutes we are seeing a request to renew the locked message a51cba32-8fd7-43c7-b4be-177099e14884, this message no longer exist because it has been already consumed.
4. Then you get this error: 'The  lock supplied is invalid. Either the lock expired, or the message has already been removed from the queue, or was received by a different receiver instance.'

Libraries are:
“Microsoft.Azure.ServiceBus 4.1.2” library with dotnet core 3.0 application

**Comment by jsquire** [comment]
//cc: @JoshLove-msft, @ShivangiReja

**Comment by nemakam** [comment]
The background timer that is used to renew locks automatically are stopped when the user returns from the messageHandler. How long do you spend within the handler? Do you do auto-complete or complete it yourself?

**Comment by gereyhe** [comment]
As per the backend the message is making the complete call in the very first seconds, however the customer has MaxAutoRenewDuration = TimeSpan.FromMinutes(300), the reason for this is because they have long running procceses. 

He has also provided me the following:

Same exceptions is encountered if message is not completed in 10 mins. Steps

1.	Subscriber pulls message
2.	Processing happens for > 10 mins
3.	Subscriber tries to unlock the message. Issues - 
a.	LockLost exception
b.	Message is not removed from queue

**Comment by nemakam** [comment]
@gereyhe , I am lost. 

> As per the backend. 

Which backend? ServiceBus backend? How do you know?

> Subscriber pulls the message

What subscriber are you talking about? Topic-Subscription? or the message-pump (`RegisterMessageHandler`) that is on the SDK? 

> Subscriber tries to unlock the message.

What is `unlock` here? Do you mean abandon?

Could you please post a sample code that explains your scenario?

**Comment by nemakam** [comment]
Also, it would be good to have the additional details I asked for - 
Is AutoComplete enabled?
Is customer completing the message himself?

**Comment by gereyhe** [comment]
Yes, Service Bus backend. I'm a Microsoft employee and I support SBUS. I have opened this Git in behalf of a customer. Regarding the subscribers, I guess the customer is talking about the consumers.

By unlock he means "completing the message (calling CompleteAsync method)"

I can't post the full code sample however I was allowed to provide some portions of it:

}
AutoComplete = false,
MaxAutoRenewDuration = TimeSpan.FromMinutes(300)
}

and later:
 
                if (!token.IsCancellationRequested && ((isSuccess.HasValue && isSuccess.Value) || !isSuccess.HasValue))
                {
                    var task = Task.Run(async () => await _queueClient.CompleteAsync(message.SystemProperties.LockToken));
                    task.Wait();
                }
                else if (token.IsCancellationRequested)
                {
                    var task = Task.Run(async () => await _queueClient.AbandonAsync(message.SystemProperties.LockToken));
                    task.Wait();
                }


And later:


        public void UnlockMessage(string lockToken)
        {
            var task = Task.Run(async () => await _queueClient.CompleteAsync(lockToken));
            task.Wait();
        }

**Comment by nemakam** [comment]
> Service Bus backend. I'm a Microsoft employee and I support SBUS.

Oh that makes sense now that you are able to see from the backend logs. So here's what I gather from the whole set..
1. Customer's handler runs for 10 minutes.
2. The moment customer receives the message, he completes them immediately.
3. Customer has set the `MaxAutoRenewDuration` as 300 minutes. 

The way the renew works in background is that for 300 minutes it will keep renewing the message OR until the handler returns. It doesn't take care of whether the message has already been completed or not by the customer. So everything that you mentioned looks like by-design. If customer's handler runs for 400 minutes, then customer will continue seeing LockLostException every 5 mins for 300 minutes (assuming the lockTimeout value on the entity is set to 5 mins).

Also, based on the code snippet you pasted, it feels like customer is calling `CompleteAsync` twice. Given that I don't have the complete code snippet, make sure customer understands that it is not valid to call this twice. The second operation will always throw LockLostException.

**Comment by ankit57** [comment]
Since IsSuccess is always returned from current code as false, the second code snippet above never gets executed. 
Let me elaborate on the issue. 

**Issue #1:** Message complete does not work if message is locked for long duration
This is the issue I started with. My process sometimes takes 10 mins to complete and the time between subscriber's first line of code and code to complete the message is > 10 mins. In this case when the code to complete message is called (snippet below), it throws MessageLockLostException. I tried increasing the MaxAutoRenewDuration to 300 minutes but that did not helped either (tried with default settings earlier by not setting value for MaxAutoRenewDuration). I had no option to proceed ahead with the development but to call complete as soon as Subscriber is called

```
public void UnlockMessage(string lockToken)
    {
        var task = Task.Run(async () => await _queueClient.CompleteAsync(lockToken));
        task.Wait();
    }
```
Note: In this case I was getting consistent error and message used to stay in bus before code retried 10 times before moving it to Dead Letter. My application is designed with assumption that message would be received once only.

**Issue #2:** Message complete after receiving but still produced MessageLockLostException

As I was blocked by Issue #1, I tried completing the message as soon as I receive it. I kept the setting for MaxAutoRenewDuration in this case. However I am getting intermittent issue where code throws MessageLockLostException however the message is completed (removed from queue). Using this I was able to proceed ahead with development however seeking resolution for Issue #1. 

Let me know if you have further question

**Comment by nemakam** [comment]
@ankit57 
Regarding the first issue - Is that a consistent repro, or happens once in a while? Why I ask this is, lets say you have the code running for 10 minutes. If either there is a communication blip in between, or if any of the VMs involved in the entire process (including ServiceBus) restarts, then you will not be able to complete the message. That's by design unfortunately. Messages can be completed on the same connection (technically AMQP link) on which it was received. 
If it is a consistent repro, then we can take a look at it and something must be wrong. As long as your handler processing time is less than `MaxAutoRenewDuration`, things should work as the SDK will renew the lock periodically.

Now coming to the second approach, the only thing it leads to is false errors. What you'd want in the case of send approach is to remove auto-renew. If you already completing the message the moment you receive it, there is no need for a background renewal. You can as well just not have a background renewal. And, given that you have requested for a background renewal, it doesn't know when you call "Complete" from within your code. It doesn't track that. The only thing it tracks is - whether you are still working in the handler or not. It stops renewing the moment your handler exits. As long as the handler is running (in this case say 10 minutes), the background task will continue to renew the message lock periodically, and every time it attempts to do that, you will end up getting a LockLostException as you have already compelted the lock and the background task doesn't know about it.

**Comment by ankit57** [comment]
The process subscribing to the message is hosted as WebJobs in webapp which has "Always On" set to "ON". The exception is not intermittent if duration for processing is more then 10 mins

Second approach was taken just to move ahead in the development. Application design assumes that message is marked as complete only when the full processing is done. I think we should focus on resolution of first issue

**Comment by nemakam** [comment]
> Regarding the first issue - Is that a consistent repro, or happens once in a while? 

To focus on first issue, could you provide this information?

**Comment by ankit57** [comment]
As mentioned in my previous comment - It is consistent

**Comment by nemakam** [comment]
Are you using the same queueClient to complete the message on which it was received?
If no, that's the problem. If yes, could you share a sample repro? This looks like a generic use-case that should be well tested.

**Comment by ankit57** [comment]
Yes. We use the same queueClient. Below is the code.

`    
public class AzureServiceBus
    {
        public event MessageQueueDelegate MessageReceivedEventHandler;

        private IQueueClient _queueClient { get; set; }
        private ILogger<AzureServiceBus> _logger;

        public MessageQueueConfigurationSettings _messageQueueConfigurationSettings { get; set; }

        public AzureServiceBus(IOptions<MessageQueueConfigurationSettings> messageQueueConfigurationSettings, ILogger<AzureServiceBus> logger)
        {
            _messageQueueConfigurationSettings = messageQueueConfigurationSettings.Value;
            _logger = logger;
        }

        public bool AddToQueue(string messageAsJson, string queueName)
        {
            _queueClient = new QueueClient(_messageQueueConfigurationSettings.ConnectionString, queueName);
            var task = Task.Run(async () => await _queueClient.SendAsync(new Message(Encoding.UTF8.GetBytes(messageAsJson))));
            task.Wait();
            return true;
        }
        public bool AddToQueue(string messageAsJson, string queueName, string connectionString)
        {
            _queueClient = new QueueClient(connectionString, queueName);
            var task = Task.Run(async () => await _queueClient.SendAsync(new Message(Encoding.UTF8.GetBytes(messageAsJson))));
            task.Wait();
            return true;
        }

        public bool CloseConnection()
        {
            var task = Task.Run(async () => await _queueClient.CloseAsync());
            task.Wait();
            return true;
        }

        public bool CreateQueue(string queueName)
        {
            return true;
        }

        public bool DeleteQueue(string queueName)
        {
            return true;
        }

        public string GetFromQueue(string queueName)
        {
            throw new NotImplementedException();
        }

        public bool IsQueueEmpty(string queueName)
        {
            return false;
        }

        public bool OpenConnection()
        {
            return true;
        }

        public void RegisterOnMessageHandlerAndReceiveMessages(string queueName)
        {
            _queueClient = new QueueClient(_messageQueueConfigurationSettings.ConnectionString, queueName);
            
            // Configure the message handler options in terms of exception handling, number of concurrent messages to deliver, etc.
            var messageHandlerOptions = new MessageHandlerOptions(ExceptionReceivedHandler)
            {
                MaxConcurrentCalls = 1,

                // Indicates whether the message pump should automatically complete the messages after returning from user callback.
                AutoComplete = false,
                MaxAutoRenewDuration = TimeSpan.FromMinutes(300)
            };

            // Register the function that processes messages.
            _queueClient.RegisterMessageHandler(ProcessMessagesAsync, messageHandlerOptions);
        }

        private async Task ProcessMessagesAsync(Message message, CancellationToken token)
        {
            try
            {
                var e = new MessageQueueMessageReceivedEventArgs(Encoding.UTF8.GetString(message.Body), message.SystemProperties.SequenceNumber,
                        message.SystemProperties.LockToken);

                bool? isSuccess = MessageReceivedEventHandler?.Invoke(this, e);
				
				// isSuccess is always false so the following code never executes
                
                if (!token.IsCancellationRequested && ((isSuccess.HasValue && isSuccess.Value) || !isSuccess.HasValue))
                {
                    var task = Task.Run(async () => await _queueClient.CompleteAsync(message.SystemProperties.LockToken));
                    task.Wait();
                }
                else if (token.IsCancellationRequested)
                {
                    var task = Task.Run(async () => await _queueClient.AbandonAsync(message.SystemProperties.LockToken));
                    task.Wait();
                }
            }
            catch (Exception ex)
            {
                // Handle exception
            }
        }

        public void UnlockMessage(string lockToken)
        {
            var task = Task.Run(async () => await _queueClient.CompleteAsync(lockToken));
            task.Wait();
        }

        public void AbandonMessage(string lockToken)
        {
            var task = Task.Run(async () => await _queueClient.AbandonAsync(lockToken));
            task.Wait();
        }
    }

`

**Comment by ankit57** [comment]
Additionally, same code works if message is completed within 30 secs. It is only when message is completed after long time (10 minutes or more)

**Comment by nemakam** [comment]
@ankit57 , are you running this on Azure? One thing that could affect this is a skew in clocks. If the service and the client are having a skew (tens of seconds), this may happen. Other than that there is no direct reason I can think of. 
We will probably have to look at it closely by monitoring the operations on service.

@gereyhe, could you work with @ankit57 (offline) and take a look at investigating issue #1. We need to understand if the client ever sent a Renew() to the service and what happened to that renew() in the backend. From the sound of it, it looks fairly easy to reproduce this issue. If just looking at backend logs doesn't help, then @ankit57 , you may have to enable client-side logs and there we can see if there is any other pattern that emerges. Once the issue is figured out, we can update this thread here with the findings (I don't want personal information about namespace etc be shared here in public).

Also, @ankit57 , do you see any other kind of exception in your exception handler?

**Comment by gereyhe** [comment]
@nemakam that we already did and that's why we opened this thread. However I cannot post the backend information into a public forum but I do have all the details. Please let me know how I can share this with you. Thanks!

**Comment by nemakam** [comment]
@gereyhe , why not do it the usual way of creating an ICM ticket internally and assigning it to ServiceBus team? For now, send me a mail to "nemakam at microsoft.com" and we can continue discussing there.

**Comment by ankit57** [comment]
@gereyhe Thanks for taking it forward. Please keep me informed about findings and resolution

**Comment by nemakam** [comment]
@aresowj, can you take a look at this one?

**Comment by danielmarbach** [comment]
Just some thoughts, @nemakam based on the snippet in https://github.com/Azure/azure-sdk-for-net/issues/11213#issuecomment-613735439 it looks like the field `queueClient` is overwritten multiple times and therefore the operations are not guaranteed to be executed on the same queue client depending on the lifetime of the instance of this class. At least the Add methods suggests it can be called multiple times.

**Comment by nemakam** [comment]
Thanks @danielmarbach , You're right. I missed that while glancing through the snippet.
@ankit57 , a message can only be completed on the same `_queueClient` it was received on. So once you recreate a queue client you will end up with a LockLostException since you are completing a message received from the previous queueClient on to the new QueueClient.
Please ensure you are calling `CompleteMessageAsync()` on the same `_queueClient` on which you did `RegisterMessageHandler`, and let us know if you still run into this problem

**Comment by nemakam** [comment]
I would also recommend you to reuse the queueClient objects. You could build a cache for queueName => queueClient objects and reuse the same client for a given queue. 
Clients are expected to be used for the entire lifetime of the process and takes care of auto-reconnect even if there is a network outage in between

**Comment by danielmarbach** [comment]
I would also recommend to avoid the sync over async. The webjobs environment has proper async entry points so that this pattern can be avoided.

**Comment by SeanFeldman** [comment]
Also, creating a `QueueClient` with a connection string upon each request is an expensive operation. I would look into caching the connection object and sharing it with the new queue clients.

**Comment by axisc** [comment]
closing this based on recommendations made on the thread. If there are still open questions, please open a new issue.
